File: implementations/fsharp/src/Skight.AgentPlatform.FSharp/TodoTool.fs
namespace Skight.AgentPlatform.FSharp

open System
open System.Collections.Generic
open System.Text
open System.Text.Json
open System.Threading.Tasks

type TodoItem = {
    Id: int
    Description: string
    mutable IsCompleted: bool
}

module TodoTool =

    let private todos = List<TodoItem>()
    let mutable private nextId = 1

    let addTodoAsync (argsJson: string) : Task<string> =
        Task.FromResult(
            lock todos (fun () ->
                try
                    let doc = JsonDocument.Parse argsJson
                    use _ = doc
                    let root = doc.RootElement
                    match root.TryGetProperty("task") with
                    | true, taskProp ->
                        let item = { Id = nextId; Description = taskProp.GetString(); IsCompleted = false }
                        nextId <- nextId + 1
                        todos.Add(item)
                        sprintf "Added TODO #%d: %s" item.Id item.Description
                    | false, _ -> "Error: Missing 'task' argument."
                with ex -> sprintf "Error adding TODO: %s" ex.Message
            )
        )

    let listTodosAsync () : Task<string> =
        Task.FromResult(
            lock todos (fun () ->
                if todos.Count = 0 then "TODO list is empty."
                else
                    let sb = StringBuilder("Current TODO List:\n")
                    for todo in todos do
                        let status = if todo.IsCompleted then "[x]" else "[ ]"
                        sb.AppendLine(sprintf "#%d %s %s" todo.Id status todo.Description) |> ignore
                    sb.ToString().TrimEnd()
            )
        )

    let completeTodoAsync (argsJson: string) : Task<string> =
        Task.FromResult(
            lock todos (fun () ->
                try
                    let doc = JsonDocument.Parse argsJson
                    use _ = doc
                    let root = doc.RootElement
                    match root.TryGetProperty("id") with
                    | true, idProp ->
                        let id = idProp.GetInt32()
                        match todos |> Seq.tryFind (fun t -> t.Id = id) with
                        | Some todo ->
                            todo.IsCompleted <- true
                            sprintf "Completed TODO #%d: %s" id todo.Description
                        | None -> sprintf "Error: TODO #%d not found." id
                    | false, _ -> "Error: Missing 'id' argument."
                with ex -> sprintf "Error completing TODO: %s" ex.Message
            )
        )

File: implementations/csharp/src/Skight.AgentPlatform/TodoTool.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;

namespace Skight.AgentPlatform
{
    public class TodoItem
    {
        public int Id { get; set; }
        public string Description { get; set; } = string.Empty;
        public bool IsCompleted { get; set; }
    }

    public static class TodoTool
    {
        private static readonly List<TodoItem> Todos = new();
        private static int _nextId = 1;

        public static Task<string> AddTodoAsync(string argsJson)
        {
            lock (Todos)
            {
                try
                {
                    using var doc = JsonDocument.Parse(argsJson);
                    var root = doc.RootElement;
                    if (!root.TryGetProperty("task", out var taskProp))
                    {
                        return Task.FromResult("Error: Missing 'task' argument.");
                    }

                    var item = new TodoItem
                    {
                        Id = _nextId++,
                        Description = taskProp.GetString()!,
                        IsCompleted = false
                    };
                    Todos.Add(item);
                    return Task.FromResult($"Added TODO #{item.Id}: {item.Description}");
                }
                catch (Exception ex)
                {
                    return Task.FromResult($"Error adding TODO: {ex.Message}");
                }
            }
        }

        public static Task<string> ListTodosAsync()
        {
            lock (Todos)
            {
                if (Todos.Count == 0) return Task.FromResult("TODO list is empty.");
                var sb = new StringBuilder("Current TODO List:\n");
                foreach (var todo in Todos)
                {
                    var status = todo.IsCompleted ? "[x]" : "[ ]";
                    sb.AppendLine($"#{todo.Id} {status} {todo.Description}");
                }
                return Task.FromResult(sb.ToString().TrimEnd());
            }
        }

        public static Task<string> CompleteTodoAsync(string argsJson)
        {
            lock (Todos)
            {
                try
                {
                    using var doc = JsonDocument.Parse(argsJson);
                    var root = doc.RootElement;
                    if (!root.TryGetProperty("id", out var idProp))
                    {
                        return Task.FromResult("Error: Missing 'id' argument.");
                    }

                    var id = idProp.GetInt32();
                    var todo = Todos.Find(t => t.Id == id);
                    if (todo == null) return Task.FromResult($"Error: TODO #{id} not found.");

                    todo.IsCompleted = true;
                    return Task.FromResult($"Completed TODO #{id}: {todo.Description}");
                }
                catch (Exception ex)
                {
                    return Task.FromResult($"Error completing TODO: {ex.Message}");
                }
            }
        }
    }
}
